### PR TITLE
address TRAC-941

### PR DESCRIPTION
**Pull request by CanOfBees**
**JIRA Ticket**: [TRAC-941](https://jira.lib.utk.edu/browse/TRAC-941)

# What does this Pull Request do?
Updates mods:name handling based on how child nodes are populated.

# What's new?
 Add a rule that will drop any mods:name that has an empty mods:namePart[@type='given' and 'family'], an empty mods:displayForm, and has a mods:role/mods:roleTerm = 'Thesis advisor' or 'Committee member'

# How should this be tested?
~~* Associate this post-processing transform
    * `vagrant destroy -f && vagrant up && vagrant ssh`
    * `git clone https://github.com/utkdigitalinitiatives/utk_isl_xml_forms`
    * `git checkout -b TRAC-941 origin/TRAC-941`
    * `sudo cp utk_isl_xml_forms/post_process_transforms/UTK_ir_etds_post_process.xsl /var/www/drupal/sites/all/modules/islandora_xml_forms/builder/self_transforms/UTK_ir_etds_post_process_941.xsl`
    * Associate the post-processing transform with the Thesis MODS Form.
        * As 'admin', use the Form Builder UI to first disable the current Thesis MODS associate and then associate this new post-process transform with the Form.
    * Pick your favorite PID and replace the MODS dsid [with this gist.](https://gist.github.com/CanOfBees/f80306ad9a2d2ef20a9c29b825724414)
    * Verify proper behavior from the post-processing transform; e.g. have empty the appropriate empty elements been dropped.~~

Edit: I wrote some incredibly poor testing instructions here. Thanks to @cdeaneGit for pointing out the problems.
### How This Should Be Considered For Testing ###
This PR presents a post-processing transform that will continue to work appropriately with data submitted through the Thesis MODS Form, so rather than testing the form submission process (you're welcome to do so, but that isn't a part of this PR) this transform should be considered as part of a separate process: will it (the stylesheet) serve to modify non-standard (and standard!) data in an external-to-Islandora transformation scenario.
* `vagrant up` and `vagrant ssh`
* Our VM doesn't have the `xsltproc` binary installed, so grab it with the following command: `sudo apt-get install xsltproc`
* Clone this repository to `/home/vagrant` and checkout the feature branch 
* cd to `utk_isl_xml_forms/`
* execute `curl -o test_data/941-test.xml https://gist.githubusercontent.com/CanOfBees/f80306ad9a2d2ef20a9c29b825724414/raw/a482fabc0604aa184941fc42e107c958413fbc25/test-mods.xml`
* execute `xsltproc -o 941-test-out.xml post_process_transforms/UTK_ir_etds_post_process.xsl test_data/941-test.xml`
* review the output to ensure validity and semantic correctness.

Alternately you can do most of that with oXygen on your desktop. You pick. :)
Apologies for bad testing instructions. Please let me know if you have questions.

## Note ##
If you try to use the test XML file in the form, it throws errors - as it should. 

## Interested Parties
@markpbaggett @cdeaneGit

### CODE REVIEW

File: post_process_transforms/UTK_ir_etds_post_process.xsl [hunk 1]
```diff
@@ -23,8 +23,17 @@
...
+    if a thesis advisor or committee member is added via the form without a namePart[@type='given' AND 'family'] AND
+    has a role/roleTerm='Thesis advisor' or 'Committee member'.
+    MODS added via the form automatically gets an empty mods:displayForm.
+  -->
+  <xsl:template match="mods:name[mods:displayForm='']
+                                [mods:namePart[@type='given']='']
+                                [mods:namePart[@type='family']='']
+                                [mods:role/mods:roleTerm='Thesis advisor' or mods:role/roleTerm='Committee member']"/>
+  <!-- if there are any empty 'type' attributes (@type), ignore them -->
+  <xsl:template match="@type[.='']"/>
```
Comment (markpbaggett) [inline post_process_transforms/UTK_ir_etds_post_process.xsl]: I like this.

File: post_process_transforms/UTK_ir_etds_post_process.xsl [hunk 2]
```diff
@@ -23,8 +23,17 @@
...
 
   <!-- if any of the following elements are empty, drop them from the transform. this list could grow. -->
-  <!-- if an extra thesis advisor or committee member is added to the form -->
-  <xsl:template match="mods:name[@type='personal'][mods:displayForm='']"/>
+  <!--
+    if a thesis advisor or committee member is added via the form without a namePart[@type='given' AND 'family'] AND
+    has a role/roleTerm='Thesis advisor' or 'Committee member'.
+    MODS added via the form automatically gets an empty mods:displayForm.
+  -->
+  <xsl:template match="mods:name[mods:displayForm='']
```
Comment (markpbaggett) [inline post_process_transforms/UTK_ir_etds_post_process.xsl:31]: We need to have a better conditional here.  If something has a blank displayForm but no nameParts, there will not be a match.

**Review by markpbaggett — CHANGES_REQUESTED** [review]

File: post_process_transforms/UTK_ir_etds_post_process.xsl (hunk 2 above)
Comment (CanOfBees) [reply post_process_transforms/UTK_ir_etds_post_process.xsl:31]: @markpbaggett @cdeaneGit please take a look at the latest changes - I think I've updated this conditional appropriately.

Comment (cdeaneGit) [reply post_process_transforms/UTK_ir_etds_post_process.xsl:31]: I started looking at this yesterday.  This was after 4, so I pulled code after you changes were in place. Still testing.

UTK_ir_etds_post_process_941.xml -- the file containing the revised code that was copied to 
            utk_ir_xml_forms/post_process_transforms/UTK_ir_etds_post_process.xsl
gist_mods.xml --the file from Bridger's gist
xsltproc   -- copied from dlwork;/bin/xsltproc and added to my vbin utility files.

I have added xsltproc to my vagrant and ran this:
>>cp gist_mods.xml orig_gist_mods.xml
>>xsltproc UTK_ir_etds_post_process_941.xml  gist_mods.xml
>>diff gist_mods.xml orig_gist_mods.xml
no diff

The gist_mods.xml is well-formed.  No problems with xmllint.
    
The PRELIMINARY_COMMENT was written earlier today, before I decided to continue testing before commenting.

[PRELIMINARY_COMMENT.txt](https://github.com/utkdigitalinitiatives/utk_isl_xml_forms/files/1492968/PRELIMINARY_COMMENT.txt)

This is a hard problem to solve.   I suggest simplifying the code by not trying to bundle roleTerm='Thesis adivisor' and roleTerm='Committee member' together.  

Also you may want to drop testing given name.  If we don't have the displayForm value and we don't have familyName, the givenName used by itself could be misleading.

Comment (CanOfBees) [reply post_process_transforms/UTK_ir_etds_post_process.xsl:31]: @cdeaneGit and @markpbaggett please review the updated testing instructions in the pull request. Let me know if you have any questions. 

Thanks!

**Comment by cdeaneGit** [comment]
I tested this following the instructions in the pull request.  All went well.
I tested the output file (941-test.xml) with xmllint on vagrant.  The file is verified as well formed.
I tested this with Oxygen on my desktop.  The file passed as valid.
I did a second test with xsltproc using a slightly different method.  The transformed file was the same as before.  
This revision took care of all the problems addressed in this ticket.

I APPROVE MERGING THIS CODE.
@CanOfBees @markpbaggett

**Review by cdeaneGit — APPROVED** [review]
I tested this following the instructions in the pull request. All went well.
I tested the output file (941-test.xml) with xmllint on vagrant. The file is verified as well formed.
I tested this with Oxygen on my desktop. The file passed as valid.
I did a second test with xsltproc using a slightly different method. The transformed file was the same as before.
This revision took care of all the problems addressed in this ticket.

I APPROVE MERGING THIS CODE.
@CanOfBees @markpbaggett

**Review by markpbaggett — APPROVED** [review]

**Pull request merged**